core/src/main/java/io/grpc/internal/ManagedChannelImpl2.java

```diff
@@ -0,0 +1,873 @@
+   * forceful, the shutdown process is still not instantaneous; {@link #isTerminated()} will likely
+   * return {@code false} immediately after this method returns.
+   */
+  @Override
+  public ManagedChannelImpl2 shutdownNow() {
+    log.log(Level.FINE, "[{0}] shutdownNow() called", getLogId());
+    shutdown();
+    delayedTransport.shutdownNow(SHUTDOWN_NOW_STATUS);
+    channelExecutor.executeLater(new Runnable() {
+        @Override
+        public void run() {
+          if (shutdownNowed) {
+            return;
+          }
+          shutdownNowed = true;
+          maybeShutdownNowSubchannels();
+        }
+      }).drain();
+    return this;
+  }
+
+  @Override
+  public boolean isShutdown() {
+    return shutdown.get();
+  }
+
+  @Override
+  public boolean awaitTermination(long timeout, TimeUnit unit) throws InterruptedException {
+    return terminatedLatch.await(timeout, unit);
+  }
+
+  @Override
+  public boolean isTerminated() {
+    return terminated;
+  }
+
+  /*
+   * Creates a new outgoing call on the channel.
+   */
+  @Override
+  public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(MethodDescriptor<ReqT, RespT> method,
+      CallOptions callOptions) {
+    return interceptorChannel.newCall(method, callOptions);
+  }
+
+  @Override
+  public String authority() {
+    return interceptorChannel.authority();
+  }
+
+  private class RealChannel extends Channel {
+    @Override
+    public <ReqT, RespT> ClientCall<ReqT, RespT> newCall(MethodDescriptor<ReqT, RespT> method,
+        CallOptions callOptions) {
+      Executor executor = callOptions.getExecutor();
+      if (executor == null) {
+        executor = ManagedChannelImpl2.this.executor;
+      }
+      StatsTraceContext statsTraceCtx = StatsTraceContext.newClientContext(
+          method.getFullMethodName(), censusFactory, stopwatchSupplier);
+      return new ClientCallImpl<ReqT, RespT>(
+          method,
+          executor,
+          callOptions,
+          statsTraceCtx,
+          transportProvider,
+          scheduledExecutor)
+              .setDecompressorRegistry(decompressorRegistry)
+              .setCompressorRegistry(compressorRegistry);
+    }
+
+    @Override
+    public String authority() {
+      String authority = nameResolver.getServiceAuthority();
+      return checkNotNull(authority, "authority");
+    }
+  }
+
+  /**
+   * Terminate the channel if termination conditions are met.
+   */
+  // Must be run from channelExecutor
+  private void maybeTerminateChannel() {
+    if (terminated) {
+      return;
+    }
+    if (shutdown.get() && subchannels.isEmpty() && oobChannels.isEmpty()) {
+      log.log(Level.FINE, "[{0}] Terminated", getLogId());
+      terminated = true;
+      terminatedLatch.countDown();
+      executorPool.returnObject(executor);
+      // Needed for delivering rejections to new calls after OobChannel is terminated.
+      executor = MoreExecutors.directExecutor();
```

Comment:
1) This is pre-existing. I would feel very comfortable leaving it as equally broken as the current code (`ManagedChannelImpl`) and fixing it later. There are so many things happening already in this PR. I'm okay with a temporary hack in order to let tests work or similar.

2) the app's executor shouldn't be called at all after termination. That applies equally to the pool, but the pool is under our control and so we have more freedom, whereas the app's executor interaction is part of externally-visible behavior. As soon as the Channel becomes terminated the application is free to shutdown their executor. Calling the listener in the same thread seems appropriate, but using `directExecutor` _here_ is missing the larger problem. The `appExecutor` is saved in `ClientCall` which is created during `newCall()`, but between `newCall()` and `ClientCall.start()` the Channel could become terminated.